File: general-solving/leetcode/0102_binary-tree-level-order-traversal.cs
/***************************************************************************
* Problem Name: Binary Tree Level Order Traversal
* Problem URL : https://leetcode.com/problems/binary-tree-level-order-traversal/
* Date        : 2015-01-05
* Complexity  : O(n), O(lg N) Stack
* Author      : Atiq Rahman
* Status      : Accepted (405ms)
* Notes       : O(lg N) Stack Space if balanced binary tree
*   Based on previous cpp solution
* meta        : tag-binary-tree, tag-bfs
***************************************************************************/
public class Solution {
  IList<IList<int>> levelList = new List<IList<int>>();

  // Using recursion/stack
  public IList<IList<int>> LevelOrder(TreeNode root, int index=0) {
    if (root == null)
      return levelList;
    if (levelList.Count <= index)
      levelList.Add(new List<int>());
    levelList[index].Add(root.val);
    LevelOrder(root.left, index + 1);
    LevelOrder(root.right, index + 1);
    return levelList;
  }

  // Using Queue
  public IList<IList<int>> LevelOrder(TreeNode root) {
    var queue = new Queue<TreeNode>( new[] { root, null });

    while (queue.Count > 0) {
      var node = queue.Dequeue();
      if (node == null) {
        if (queue.Count == 0) continue;
        levelList.Add(new List<int>());
        queue.Enqueue(null);
      }
      else {
        levelList[levelList.Count-1].Add(node.val);
        queue.Enqueue(node.left);
        queue.Enqueue(node.right);
      }
    }
    return levelList;
  }
}

File: general-solving/leetcode/0102_binary-tree-level-order-traversal.java
/***************************************************************************************************
* Title : Binary Tree Level Order Traversal
* URL   : https://leetcode.com/problems/binary-tree-level-order-traversal/
* Date  : 2019-03
* Comp  : O(n), O(lg N)
* Author: Atiq Rahman
* Status: Accepted
* Notes : O(lg N) Stack Space if balanced binary tree
*   Based on previous cpp solution
*   Please note that this is problem is solved in cpp and java
* rel   : 'leetcode 107 Binary Tree Level Order Traversal II'
* meta  : tag-ds-binary-tree, tag-recursion, tag-graph-bfs, tag-leetcode-medium
***************************************************************************************************/
class Solution {
  List<List<Integer>> levels;
  
  public List<List<Integer>> levelOrder(TreeNode root) {
    levels = new ArrayList<List<Integer>>();
    levelOrder(root, 0);
    return levels;
  }
                      
  private void levelOrder(TreeNode root, int depth) {
    if (root == null)
      return ;
    
    if (levels.size() == depth)
      levels.add(new LinkedList<Integer>());

    levels.get(depth).add(root.val);
    levelOrder(root.left, depth + 1);
    levelOrder(root.right, depth + 1);
  }
}
